### PR TITLE
=str Add LazyFutureSource to avoid sub materialization.

### PR DESCRIPTION
**Pull request by He-Pin**
refs: https://github.com/apache/incubator-pekko/issues/46

### CODE REVIEW

**Comment by mdedetrich** [comment]
@He-Pin You need to rebase this

**Comment by mdedetrich** [comment]
@He-Pin Do you want to look into this now that we are on 1.1.x series (i.e. rebase on current main)? Also if possible and easy to do so would it be possible to add a benchmark in https://github.com/apache/incubator-pekko/tree/b0fdac259bd57fdd481483f3fe9a7aec6e1ff38a/bench-jmh/src/main/scala/org/apache/pekko/stream ? While I am sure that you version has better performance, the adding of the benchmark is to prevent any potential future performance regressions.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (mdedetrich) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala:1]: This header file is a slightly different variant compared to what else is in Pekko, see https://github.com/apache/incubator-pekko/blob/bd9ff4da5bb59234c39af8916fb7a06caa755052/project/AddMetaInfLicenseFiles.scala#L1-L16 for the correct version

**Comment by He-Pin** [comment]
@mdedetrich I will find some time to update these PR, but will afterwork, busy days, maybe this weekend.

**Comment by mdedetrich** [comment]
No worries its not urgent

File: stream/src/main/scala/org/apache/pekko/stream/scaladsl/Source.scala [hunk 2]
```diff
@@ -13,16 +13,15 @@ import scala.compat.java8.FutureConverters._
 import scala.concurrent.{ Future, Promise }
 import scala.concurrent.duration.FiniteDuration
 
-import org.reactivestreams.{ Publisher, Subscriber }
-
 import org.apache.pekko
+import org.reactivestreams.{ Publisher, Subscriber }
```
[the remark this answers is not in the document]
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/scaladsl/Source.scala:17]: @pjfanning I will do a sortImports when I back to this PR.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * license agreements; and to You under the Apache License, version 2.0:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * This file is part of the Apache Pekko project, which was derived from Akka.
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala:7]: I think this is the wrong header.

@He-Pin did you write this code yourself or did you copy it from Akka. If you wrote it and did an Akka PR, you still own the copyright on the code.

I think the code should have an ASL header like this file:

https://github.com/apache/incubator-pekko/blob/main/project/AddMetaInfLicenseFiles.scala

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+package org.apache.pekko.stream.impl.fusing
+import scala.concurrent.Future
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala:11]: can you leave a blank line before the imports?

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala (hunk 3 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala:7]: I copied  it from another file, will update.

Comment (pjfanning) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/LazyFutureSource.scala]: that special header saying `derived from Akka` is only for code that we took from Akka

**Review by pjfanning — APPROVED** [review]
lgtm

**Review by pjfanning — APPROVED** [review]
lgtm

File: .scalafix.conf [hunk 5]
```diff
@@ -47,4 +47,5 @@ SortImports.blocks = [
   "*",
   "com.sun."
   "org.apache.pekko."
+  "org.reactivestreams."
 ]
```
Comment (mdedetrich) [inline .scalafix.conf:51]: nit: Can you add a newline here

**Review by mdedetrich — APPROVED** [review]
Looks good with me, as I said in https://github.com/apache/incubator-pekko/pull/48#issuecomment-1657805114 it would be nice if a simple benchmark was added but if you don't want to do that now you can just create an issue and we can do it later.

File: .scalafix.conf (hunk 5 above)
Comment (mdedetrich) [inline .scalafix.conf:51]: nit: Can you add a newline here

**Pull request merged**